Added Status Moves

SandytheLizard opened the pull request:
Allows status type moves in technique JSONs

SandytheLizard committed:
Flying Type and Overfeed Move/Effect

Adds a flying type, nothing has the type yet, its just a placeholder.
Adds the move overfeed that heals the hp of the target to full but halves their speed.

SandytheLizard committed:
Fixed Accidently deleting lifeleech from translations

SandytheLizard committed:
Update technique.py

Fixed the return failed

SandytheLizard committed:
Update formula.py

deleted flying type for now

SandytheLizard committed:
Update technique.py

forgot something

SandytheLizard committed:
Allow Stat Chanegs in Move JSON [WIP]

Working on making stat changes work in the JSON move files. Almost done

SandytheLizard committed:
Merge branch 'development' of https://github.com/SandytheLizard/Tuxemon into development

SandytheLizard committed:
Status Effect Moves FInished

Allows For Status based moves in JSON Techniques

bitcraft commented on tuxemon/technique.py:
```diff
@@ -238,35 +254,63 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+    def changestats(self, user: Monster, target: Monster) -> EffectResult:
+        statsmaster = [self.statspeed, self.stathp, self.statarmour, self.statmelee, self.statranged, self.statdodge]
+        statslugs = ['speed', 'hp', 'armour', 'melee', 'ranged', 'dodge']
+        print(statsmaster)
```
please remove this print statement

bitcraft commented on tuxemon/technique.py:
```diff
@@ -238,35 +254,63 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+        print(statsmaster)
+        if None not in statsmaster and [] not in statsmaster:
+            for stat, slug in zip(statsmaster, statslugs):
+                print(stat)
```
print

bitcraft commented on tuxemon/technique.py:
```diff
@@ -238,35 +254,63 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+                override = stat['overridetofull']
+                basestatvalue = getattr(target, slug)
+                if value > 0 and override == False:
+                    print(basestatvalue)
```
print

bitcraft commented on tuxemon/technique.py:
```diff
@@ -238,35 +254,63 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+                        newstatvalue = basestatvalue + value
+                    else:
+                        newstatvalue = basestatvalue // value
+                    print(newstatvalue)
```
print

bitcraft commented on mods/tuxemon/db/technique/overfeed.json:
```diff
@@ -0,0 +1,53 @@
+  "potency": 1,
+  "power": 0,
+  "statspeed": {
+    "value": 2, "dividing":true, "overridetofull":false
```
lets document these values somewhere.  im not familiar with this, and its not clear what "dividing", and "overridetofull" mean.  my suggestion would be to add the documentation to the function which uses these values.

SandytheLizard replied:
ok, ill do that in the technique.py

bitcraft commented on tuxemon/technique.py:
```diff
@@ -76,17 +75,17 @@ class Technique:
         self.accuracy = 0.0
         self.animation = ""
-        self.can_apply_status = False
+        self.can_apply_status = True
```
Why was the default changed?

SandytheLizard replied:
it was causing problems with the original overfeed code, i changed it back as that code no longer exists.

bitcraft commented on tuxemon/technique.py:
```diff
@@ -99,6 +98,12 @@ def __init__(
         self.next_use = 0.0
         self.potency = 0.0
         self.power = 1.0
+        self.statspeed: Sequence[str] = []
```
The typing is incorrect.  The JSON defines these values as a mapping/dict, not Sequence

SandytheLizard replied:
Not quite sure what you mean here, what would should I change it too?

vnmabus replied:
The thing is that in the JSON you have defined them to be a dictionary (a key-value mapping) and not a sequence.

In order to correctly type it you should create a typed dict class in `db.py`:
```python
class JSONStat(TypedDict):
    value: int  # I am not sure if this is supposed to be int or float, you decide
    dividing: bool
    overridetofull: bool
```
and use that to type the stat fields in `JSONTechniqueOptionalFields` and here.

SandytheLizard replied:
fixed it, but since that file isnt in this PR ill upload it in the next one

vnmabus replied:
The file IS in this PR, and you have already modified it. But you have repeated the typeddict definition in technique.py, try importing it from db.py.

bitcraft commented on tuxemon/technique.py:
```diff
@@ -238,35 +254,63 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+        statsmaster = [self.statspeed, self.stathp, self.statarmour, self.statmelee, self.statranged, self.statdodge]
+        statslugs = ['speed', 'hp', 'armour', 'melee', 'ranged', 'dodge']
+        print(statsmaster)
+        if None not in statsmaster and [] not in statsmaster:
```
This check is difficult to understand.  i think that you are looking for an empty list, or a list of Nones.   Consider rewriting it by iterating over the zippled list first, then checking for None or empty values.  It would be easier to understand what is going on.

SandytheLizard replied:
im unsure how to do that. The reason i did this is sometimes it returns the following:
[[],[],[].[],[],[]] or [None, None, None, None, None, None] which it then gives a nonetype error and crashes the program.

bitcraft replied:
please don't resolve comments if there are no changes to address it.  try something like this:
```python
for stat, slug in zip(statsmaster, statslugs):
    if not stat:
        continue
```

bitcraft commented on tuxemon/technique.py:
```diff
@@ -309,6 +353,7 @@ def damage(self, user: Monster, target: Monster) -> EffectResult:
             if not hit:
                 damage //= 2
             target.current_hp -= damage
+            success = damage
```
this would probably be better as
```python
success = damage > 0
```
in general, passing numeric values around and treating the as booleans can lead to bugs or confusing code later.  better to cast as the correct type as soon as possible

SandytheLizard replied:
ill just delete that, i added it and its not needed anymore anyway

bitcraft commented on tuxemon/technique.py:
```diff
@@ -332,11 +377,12 @@ def apply_status(self, slug: str, target: Monster) -> EffectResult:
+        success = not already_applied and self.potency >= random.random()
         tech = None
         if success:
-            tech = Technique(slug, carrier=target)
```
why was the kwarg removed?

SandytheLizard replied:
accident, ill fix it

bitcraft commented on tuxemon/technique.py:
```diff
@@ -356,7 +402,7 @@ def apply_lifeleech(self, user: Monster, target: Monster) -> EffectResult:
 
         """
         already_applied = any(t for t in target.status if t.slug == "status_lifeleech")
-        success = not already_applied and self.can_apply_status and self.potency >= random.random()
```
why was can_apply_status removed?

SandytheLizard replied:
as i said above, it was casuing issues with code that no longer exists, ill change it back

bitcraft replied:
if its causing problems, then maybe it can be removed, i just want to make sure there is a reason

bitcraft commented on tuxemon/technique.py:
```diff
@@ -366,7 +412,6 @@ def apply_lifeleech(self, user: Monster, target: Monster) -> EffectResult:
         }
 
     def poison(self, target: Monster) -> EffectResult:
-        assert self.link
```
assert statements are useful for debugging, or catching errors at runtime.  im not familiar with this one, but why was it removed?

SandytheLizard replied:
was causing crashes

bitcraft reviewed (changes requested):
lots of small comments, but overall looks good.  thanks for the pr and im looking forward to merging it

bitcraft commented:
Also note there are merge conflicts, so please pull in the latest changes and merge if needed to fix that.

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

Removed Print Statements

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Update technique.py

SandytheLizard committed:
Merge branch 'development' into development

SandytheLizard committed:
Levels and Stats

Fixed Tuxemon using moves of a too high level, also added some status stuff to DB.py  as requested.

SandytheLizard reviewed (commented):
Fixed Everything , also added the level fixes

bitcraft commented on tuxemon/technique.py:
```diff
@@ -71,12 +70,15 @@ def merge_results(result: EffectResult, meta_result: TechniqueResult) -> Techniq
     meta_result.update(result)
     return meta_result
 
+class JSONStat(TypedDict):
```
instead of duplicating this, please import it from the db, if possible

SandytheLizard committed:
Update technique.py

bitcraft commented on mods/tuxemon/l18n/en_US/LC_MESSAGES/base.po:
```diff
@@ -170,7 +170,9 @@ msgid "combat_state_lifeleech_get"
 
 msgid "combat_state_overfeed_get"
-msgstr "{user} Overfed {target}."
+
```
just a comment, no need to change this, but please avoid changing indentation/white space in files unless you are correcting formatting.  in this case, an extra newline was added.

bitcraft commented on tuxemon/technique.py:
```diff
@@ -155,6 +164,15 @@ def load(self, slug: str) -> None:
             self.type1 = self.type2 = None
 
         self.power = results.get("power", self.power)
+
```
extra newline added

bitcraft commented on tuxemon/technique.py:
```diff
@@ -202,6 +220,7 @@ def reset_combat_counter(self) -> None:
         self._combat_counter = 0
 
     def use(self, user: Monster, target: Monster) -> TechniqueResult:
+        
```
extra newline

bitcraft reviewed (changes requested):
formatting changes i can live with, but please check the other comments

SandytheLizard committed:
Update technique.py

SandytheLizard reviewed (commented):
Simplified the stat check code. Also imported db

bitcraft commented on tuxemon/technique.py:
```diff
@@ -272,7 +287,34 @@ def use(self, user: Monster, target: Monster) -> TechniqueResult:
+        statsmaster = [self.statspeed, self.stathp, self.statarmour, self.statmelee, self.statranged, self.statdodge]
+        statslugs = ['speed', 'hp', 'armour', 'melee', 'ranged', 'dodge']
+        for stat, slug in zip(statsmaster, statslugs):
+            if not stat:
```
this is a bug.  my previous suggestion was
```python
if not stat:
    continue
```
right now, if stat is None or an empty list, there will be an error when it is used as a dictionary.  please fix this and play test the game.

SandytheLizard replied:
Ok, sorry thought the continue was a placeholder. I'll fix that.

bitcraft reviewed (changes requested):
see comments

SandytheLizard committed:
Update technique.py

Added Continue

SandytheLizard committed:
Merge branch 'development' of https://github.com/SandytheLizard/Tuxemon into development

SandytheLizard reviewed (commented):
should be good now
